activate.go

```diff
@@ -0,0 +1,665 @@
+	// and then adding them back when the rate limiting expires. Right now, any change
+	// to authType requires us to restart systemd-ask-password (assuming the systemd
+	// implementation of AuthRequestor is used), which won't be a great experience if,
+	// eg, the user is in the middle of entering a recovery key having just entered an
+	// incorred PIN, when PIN becomes available again, requiring us to add back
+	// UserAuthTypePIN to authType and restarting systemd-ask-password in order to
+	// update the prompt - the result will be that the update to the prompt will cause
+	// the user to lose what they've entered so far.
+	//
+	// Maybe rate limiting will require us to replace systemd-ask-password with
+	// something that fits this user experience better in the future.
+
+	// Determine the available authentication types.
+	var authType UserAuthType
+	if len(recoverySlotRecords) > 0 {
+		authType |= UserAuthTypeRecoveryKey
+	}
+	// TODO: Update authType for PIN / passphrase
+
+	// XXX: When passphrase + PIN support lands, this will loop on available
+	// PIN + passphrase tries as well.
+	for recoveryKeyTries > 0 {
+		// Update authType flags
+		// XXX: This code will make more sense once support for passphrases and PINs is in,
+		// as what we're doing here is removing an auth type once the number of tries has
+		// expired so that the user prompt can be updated.
+		if recoveryKeyTries == 0 {
+			// No more recovery key tries are left,
+			authType &^= UserAuthTypeRecoveryKey
+		}
+
+		cred, err := authRequestor.RequestUserCredential(ctx, name, m.container.Path(), authType)
+		if err != nil {
+			return fmt.Errorf("cannot request user credential: %w", err)
+		}
+
+		// We have a user credential.
+		// 1) TODO: Try it against every keyslot with a passphrase.
+		// 2) TODO: See if it decodes as a PIN and try it against every keyslot with a passphrase.
+		// 3) See if it decodes as a recovery key, and try it against every recovery keyslot.
+
+		recoveryKey, err := ParseRecoveryKey(cred)
+		if err == nil {
+			// This is a valid recovery key
+			recoveryKeyTries -= 1
+			if slot := m.tryRecoveryKeyslots(ctx, recoverySlotRecords, recoveryKey); slot != nil {
+				// Success!
+				m.status = ActivationSucceededWithRecoveryKey
+			}
+
+		}
+
+		if m.status == ActivationIncomplete {
+			// We haven't unlocked yet, so try again.
+			continue
+		}
+
+		// We have unlocked successfully.
+		m.next = activateOneContainerStateMachineTask{
+			name: "add-keyring-keys",
+			fn: func(ctx context.Context) error {
+				return m.addKeyringKeys(ctx, DiskUnlockKey(recoveryKey[:]), PrimaryKey(nil))
+			},
+		}
+		return nil
+	}
+
+	// We have failed to unlock
+	m.next = activateOneContainerStateMachineTask{}
+
+	return ErrCannotActivate
+}
+
+func (m *activateOneContainerStateMachine) tryRecoveryKeyslots(ctx context.Context, slotRecords keyslotAttemptRecordSlice, recoveryKey RecoveryKey) Keyslot {
```

Comment:
this is a helper and not one of the states, not sure if we should mark it somehow but it got me slightly confused